[exir][delegate] torch.ops.call_delegate

angelayi opened the pull request:
Summary:
Followup diffs to integrate this op into the other parts of the delegate workflow.
The unittest results in the following graph:

```
graph():
    %x_1 : [#users=1] = placeholder[target=x_1]
    %y_1 : [#users=1] = placeholder[target=y_1]
    %lowered_module_0 : [#users=1] = get_attr[target=lowered_module_0]
    %call_delegate : [#users=1] = call_function[target=torch.ops.call_delegate](args = (%lowered_module_0, forward, %x_1, %y_1), kwargs = {})
    return call_delegate
```

Test Plan: buck2 run //executorch/exir/tests:delegate -- -r "test_call_delegate"

Differential Revision: D42329287



cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

voznesenskym commented on torch/_dynamo/variables/torch.py:
```diff
@@ -832,6 +832,20 @@ def speculate_branch(branch):
             # TODO: assert that the true/false return values are
             # consistent
             example_value = true_r.as_proxy().node.meta["example_value"]
+        elif self.value.__name__ == "call_delegate":
```
if call_delegate is from executorch/exir - I am not sure we want to reference it by this string notion here... where is def call_delegate?

raziel commented on torch/_dynamo/variables/torch.py:
```diff
@@ -900,6 +901,26 @@ def speculate_branch(branch):
             example_value = r.new_empty(
                 [get_fake_value(args[1].as_proxy().node, tx).shape[0], *r.shape]
             )
+        elif self.value.__name__ == "executorch_call_delegate":
```
Hi @voznesenskym 
I imagine this won't be the long term approach to handle these special opaque nodes. Meaning, there would need to be a pluggable way for dynamo to take them in without having to hardcode them.

Could you please confirm?

voznesenskym replied:
Your imagination serves you well, as always. However, we do not have any design yet - I called this out in another chat (you maybe saw) - atm, there are some hierarchical violations. 

They are not special opaque nodes, but they are python operators, registered with the pyop system.  Ops are lower level than dynamo, yet dynamo knows how to trace them. The biggest issue is actually that this is a non core op, and is user provided, yet dynamo knows about it. 

The violation therefore, is not in the dynamo<>pyoperator relationship, which is basically the same as dynamo<>any torch op relationship - but in the relationship of dynamo (and the whole torch stack) with custom ops. And this is not a new discussion, fwiw, this has always been a tension.

raziel replied:
Sorry I haven't caught up with threads today.
I'm reviewing other of @angelayi 's changes (internal exir/executorch) and was curious whether there was a resolution on this, so I figured ask directly here in the code: I just wanted to get a sense of where your thinking was at the moment in terms of the design, and a bit timeline (I shall look at the other thread!).
Thanks!

cccclai replied:
> this is a non core op, and is user provided

Just would like to clarify, this op is not user provided, and it's an interface we provide for users to interact with backends. What do you consider is core op, other than cond?

cccclai replied:
> a pluggable way for dynamo to take opaque nodes

Is there any ongoing discussion about it? I understand that ideally, we don't want to do too many special cases with the custom op, however, I do think the interface with backends is generic and is a fundamental capability. It can also be applied in eager mode such that we can interact with different backends smoothly.

voznesenskym reviewed (approved):
if CI is happy, so am I, godspeed

voznesenskym commented on torch/_dynamo/variables/torch.py:
```diff
@@ -900,6 +901,26 @@ def speculate_branch(branch):
+            )
+
+            fake_sub_args = tuple(arg.get_real_value() for arg in args[2:])
+            example_real_value = getattr(lowered_module.original_module, func_name)(
+                *fake_sub_args
+            )
```
we may regret this in the future, consider executing it without dispatches...

angelayi committed:
[exir][delegate] torch.ops.call_delegate (#92562)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/92562

Context: https://docs.google.com/document/d/1KMHuGwDPROqIa0nsweY5iGOJIcgWFjirtgY7UJMmgLE/edit

Followup diffs to integrate this op into the other parts of the delegate workflow.
The unittest results in the following graph:

```
graph():
    %x_1 : [#users=1] = placeholder[target=x_1]
    %y_1 : [#users=1] = placeholder[target=y_1]
    %lowered_module_0 : [#users=1] = get_attr[target=lowered_module_0]
    %call_delegate : [#users=1] = call_function[target=torch.ops.call_delegate](args = (%lowered_module_0, forward, %x_1, %y_1), kwargs = {})
    return call_delegate
```

Test Plan: buck2 run //executorch/exir/tests:delegate -- -r "test_call_delegate"

Reviewed By: voznesenskym, tugsbayasgalan

Differential Revision: D42329287

fbshipit-source-id: d000ec82d89e384df263af02630a74be802378ed
